content/integrations/google_cloud/google_cloud_pubsub/Java.md

```diff
@@ -0,0 +1,429 @@
+     try (Scope traceScope =
+            OpenCensusTraceUtil.createScopedSampledSpan(
+               "PublisherRoot-" + i)) {
+       OpenCensusTraceUtil.addAnnotation(
+         OpenCensusTraceUtil.getCurrentSpanIdAsString());
+       String message = "message-" + i;
+       // convert message to bytes
+       ByteString data = ByteString.copyFromUtf8(message);
+       PubsubMessage pubsubMessage = PubsubMessage.newBuilder()
+         .setData(data)
+         .build();
+       // Schedule a message to be published. Messages are
+       // automatically batched.
+       ApiFuture<String> future = publisher.publish(pubsubMessage);
+       futures.add(future);
+     }
+   }
+ } finally {
+    // Wait on any pending requests
+   List<String> messageIds = ApiFutures.allAsList(futures).get();
+   for (String messageId : messageIds) {
+     System.out.println(messageId);
+   }
+   if (publisher != null) {
+     // When finished with the publisher, shutdown to free
+     // up resources.
+     publisher.shutdown();
+   }
+   OpenCensusTraceUtil.addAnnotation("Publisher:End");
+   Thread.sleep(5000);
+ }
+}
+```
+
+The most important change in the code is adding:
+
+```java
+.setTransform(OpenCensusUtil.OPENCENSUS_MESSAGE_TRANSFORM)
+```
+
+when building the Publisher, which enables the implicit propagation of the
+current OpenCensus trace span when publishing a message. Strictly speaking, this
+code change is the only required change in the Publisher code. The other changes
+add OpenCensus instrumentation to the application to give more insight about
+what’s going on during execution.
+
+The first code change is in the outer try-with-resources block:
+
+```java
+try (Scope scope =
+      OpenCensusTraceUtil.createScopedSampledSpan("Publisher")) ...
+```
+
+which creates a top-level parent span named ‘Publisher’ that is live for the
+execution of the application (`OpenCensusTraceUtil` is a simple utility class
+added to the example  —  details can be seen in the
+[Appendix](#appendix)). Because only sampled spans are propagated, for demo
+purposes, we created a sampled parent span. However in standard applications,
+please use the default sampling rate.
+
+The next code change:
+
+```java
+OpenCensusTraceUtil.addAnnotation("Publisher:Begin");
+```
+
+adds an annotation to the parent span at the beginning of execution, before
+sending any messages (we’ll see this exported later).
+
+The third code change: 
+
+```java
+try (Scope traceScope = OpenCensusTraceUtil.createScopedSampledSpan(
+      "PublisherRoot-" + i)) {
+```      
+
+adds a child span to our parent span for each message published — the child span
+is named PublisherRoot-i to make it easy to identify in the trace.
+
+The next change is to add an annotation with the child span's span id, which
+allows us to connect publisher spans and subscribers spans in the next section.
+
+```java
+OpenCensusTraceUtil.addAnnotation(
+         OpenCensusTraceUtil.getCurrentSpanIdAsString());
```

Comment:
It would be great if we just used the actual APIs for creating links otherwise we are introducing magic that is unexplained despite this whole article -- the gist of it is asynchronous processing and using links but unfortunately I don't see any of https://opencensus.io/tracing/span/link/#1 shown